### PR TITLE
[3/n] Fix the bug where Python mode fails to dispatch detach()

### PR DESCRIPTION
**Pull request by cbalioglu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68965
* #68955
* __->__ #68954
* #68953
* #68952

This PR fixes a subtle bug where the `detach()` operation of tensors with no `__torch_dispatch__` are not dispatched in Python mode.

Differential Revision: [D32679739](https://our.internmc.facebook.com/intern/diff/D32679739/)

cc @albanD @zou3519

### CODE REVIEW

File: test/test_python_dispatch.py [hunk 1]
```diff
@@ -447,5 +447,24 @@ def wrap(e):
...
+                if func == torch.ops.aten.detach:
+                    num_detach_dispatched += 1
+
+                with no_dispatch():
+                    return func(*args, **kwargs)
+
+        with enable_python_mode(FakeTensor):
+            tmp = torch.ones([10, 10]).detach()
+
+        self.assertEqual(num_detach_dispatched, 2)
```
Comment (zou3519) [inline test/test_python_dispatch.py:467]: @albanD I always thought it was weird that Tensor.detach() shows up in `__torch_dispatch__`. If we are marketing `__torch_dispatch__` as allowing users to override "the PyTorch API surface after autograd", then shouldn't detach not be a part of that API?

Comment (albanD) [reply test/test_python_dispatch.py:467]: I was having this exact discussion with Jeffrey on the forward AD functions like make/unpack dual.
I think that once we have pre dispatch extension they should not indeed. The current implementation is hacky to be able to catch it even though it shouldn't show up.

Note that the way to do this is simply to implement the detach function without a redispatch at the autograd level.

**Comment by albanD** [comment]
Do you have more details on the subtle bug you mentioned above and why this is the right way to fix it?

**Comment by cbalioglu** [comment]
> Do you have more details on the subtle bug you mentioned above and why this is the right way to fix it?

Hey @albanD, the subtle bug I mentioned is as an end user I would expect `__torch_dispatch__` to catch `detach()` calls like other Torch operators since it is ultimately an operator as well. Although it is handled specially due to Autograd machinery, I would expect that to be an internal implementation detail.

**Comment by albanD** [comment]
Interestingly, as @zou3519 mentioned above, it is most likely the other way around. You should never see dispatch when you are "below" autograd. Or maybe you should only see a call to "alias"

**Comment by cbalioglu** [comment]
I see. Is there any other mechanism (either in Python or C++) to hook custom code to the detach operation?

Just to give a bit more context, in my work to materialize parts of a model (i.e. submodules or specific tensors), the `meta_init` context manager uses a simple operator tracing. The following code snippet shows its low-level API:

```
>>> with torch.distributed.nn.utils.meta_init():
...     t1 = torch.ones([3])
...     t2 = t1 + 2
>>> t1
tensor(..., device='meta', size=(3,))
>>> t2
tensor(..., device='meta', size=(3,))
>>> t1 = torch.distributed.nn.utils.materialize_tensor(t1, keep_meta_cache=True)
>>> t1
tensor([1., 1., 1.])
>>> t2 = torch.distributed.nn.utils.materialize_tensor(t2)
>>> t2
tensor([3., 3., 3.])
```
The implementation handles all intricacies of in-place updates, shared storage views, etc. and only materializes the operation nodes that are necessary for the final output requested by `materialize_tensor()`. You can check out the WIP implementation [here](https://github.com/cbalioglu/pytorch/blob/no-init-2/torch/distributed/nn/utils/meta_init.py). However since `detach()` is a valid operator that the user can call like any other operator, I have trouble building the graph if I cannot "trace" it like other operators. That was also my main motivation for this PR. So if you can suggest any other workaround, I would appreciate it.

**Comment by albanD** [comment]
Is your materialization supposed to be differentiable in the same way as the original code?
If so you would need to "trace" above autograd (and so using a pre-dispatch kind of key or torch function).

If you trace below, autograd, then you will only see the ops that are actually computing the result.

**Comment by zou3519** [comment]
@cbalioglu it sounds like what you really want here is a `__torch_function__` mode that works like our PythonMode (which is a `__torch_dispatch__` mode.

`__torch_function__` lets you override operations at the Python level. `__torch_dispatch__` lets you override operations that occur after autograd. As Alban pointed out, detach() should not be recorded in `__torch_dispatch__` because detach is something that happens before autograd.

The downside to `__torch_function__` is that there is currently no way to override factory functions (i.e. functions that take in no tensor arguments like `torch.empty` ctor, so we'd want a "`__torch_function__` mode".

What do you think?

**Comment by cbalioglu** [comment]
Thanks for the feedback @zou3519, @albanD . The problem I see with `__torch_function__` (correct me if I am wrong) is that you cannot have a global catch-all `__torch_function__` similar to what you achieve with the Python Mode + `__torch_dispatch__` combination. Your `__torch_function__` gets called only if you have a tensor subclass instance and that instance is passed as an argument to an ATen operator (and no other argument before you in the lexical order claims to handle the operator).

As you are aware, in the meantime I have started implementing a "MetaInit" backend in C++ that fulfills these requirements. Probably for this discussion the most relevant PR is #69637.

**Comment by cbalioglu** [comment]
Closing this PR, thanks again for your feedbacks!
